classification_analysis.ipynb

```diff
@@ -1107,22 +1141,41 @@
     "        df = df.reindex(\n",
     "            columns=['WALKING', 'BICYCLING', 'E_BIKE', 'ESCOOTER', 'CAR', 'BUS', 'SUBWAY', 'LIGHT_RAIL', 'TRAIN', 'NO_GT_START', 'NO_GT_MIDDLE', 'NO_GT_END']\n",
     "        ).fillna(0)\n",
-    "#         df = df.div(df.sum(axis=1), axis=0)\n",
+    "\n",
+    "        if normalization == 'pred':\n",
+    "            # After transposing, predictions are axis=0\n",
+    "            df = df/df.sum(axis=1, skipna=True)\n",
+    "        elif normalization == 'gt':\n",
+    "            # After transposing, GTs aree axis=1\n",
+    "            df = df/df.sum(axis=0, skipna=True)\n",
+    "        elif normalization == 'all':\n",
+    "            # Axis-agnostic.\n",
+    "            df = df/df.sum(skipna=True)\n",
+    "        \n",
+    "        # div-by-zero causes NaN.\n",
+    "        df = df.fillna(0)\n",
+    "\n",
     "        cm = ax[k].imshow(df.transpose(), interpolation='nearest',  cmap=plt.cm.coolwarm, aspect='auto')\n",
     "        ax[k].set_title(role)\n",
-    "    #     plt.colorbar(cm, ax=ax[0])\n",
     "        tick_marks = np.arange(len(df))\n",
     "        ax[k].set_yticks(np.arange(len(df.columns)))\n",
     "        ax[k].set_xticks(np.arange(len(df)))\n",
     "        ax[k].set_yticklabels(df)\n",
     "        ax[k].set_xticklabels(df.index, rotation=80)\n",
     "        color_thresh = df.max().max() / 4\n",
     "        for i, j in itertools.product(range(df.shape[1]), range(df.shape[0])  ):\n",
-    "            ax[k].text(j, i, (int(df.transpose().iat[i,j])), horizontalalignment='center', \n",
-    "                   color='white' \n",
-    "                       if df.transpose().iat[i,j] < color_thresh \n",
-    "#                            or df.transpose().iat[i,j] in df.max()\n",
-    "                       else 'black')\n",
+    "\n",
+    "            if normalization is None:\n",
+    "                ax[k].text(j, i, (int(df.transpose().iat[i,j])), horizontalalignment='center', \n",
+    "                       color='white' \n",
+    "                           if df.transpose().iat[i,j] < color_thresh \n",
+    "                           else 'black')\n",
+    "            else:\n",
+    "                ax[k].text(j, i, np.round(df.transpose().iat[i,j], 3), horizontalalignment='center', \n",
+    "                    color='white' \n",
+    "                        if df.transpose().iat[i,j] < color_thresh \n",
+    "                        else 'black')\n",
+    "                        \n",
```

Comment:
what is the rationale for this change (e.g. adding the `3` parameter?)

Reply:
After normalizing, the calculated values are floating point values and do not round-off whilst displaying the confusion matrix. This makes the display texts overlap each other, rendering the entire confusion matrix unreadable. If normalization is not required, we cast to int (default implementation)

Reply:
I assumed that 3 digit precision would be an ideal trade-off between readability and precision.